### ISSUE TITLE
Feature request - port to Python 3.6

### ISSUE THREAD

**Issue by znmeb** [plaidml/plaidml#20]
I've got PlaidML running on my AMD Bonaire on Arch Linux with Python 2.7 in a Conda environment. Every other Python package I have runs with 3.6 and my goal is to keep it that way. ;-)

There doesn't seem to even be a pip package for 3.6, so the `pip install -U plaidml-keras` fails with Python 3.6. If you can post build-from-GitHub-source instructions, I can make a local package and install it.

P.S.: Let me know if you want Arch setup instructions for AMD GPUs. Most of it is on the Arch User Repository wiki but I've got some scripts that do the work.

P.P.S.: Benchmark results
```
Using PlaidML backend.
INFO:plaidml:Initializing device bonaire.0: "Bonaire", vendor "Advanced Micro Devices, Inc."
INFO:plaidml:Opening device "bonaire.0": "Advanced Micro Devices, Inc. Bonaire"
Downloading data from https://github.com/fchollet/deep-learning-models/releases/download/v0.6/mobilenet_1_0_224_tf.h5
16793600/17225924 [============================>.] - ETA: 0s 
Model loaded.
Compiling and running initial batch, batch_size=1
Warmup
Doing the main timing
Example finished, elapsed: 6.821215868 (compile), 15.0223557949 (execution)
```

**Comment by brianretford** [comment]
We'll add build instructions tonight. Until then:
Install bazel 6.1
bazel build plaidml:wheel plaidml/keras:wheel
sudo pip install bazel-bin/plaidml/*whl bazel-bin/plaidml/keras/*whl

(Wrote this on my phone, haven't tested, looks right to me though)

**Comment by znmeb** [comment]
Thanks! Does it require Bazel 6.1? I have an older Bazel for building TensorFlow.

**Comment by brianretford** [comment]
Maybe? It does for our internal stuff, not sure about just the open source
part. Run it and let us know

**Comment by znmeb** [comment]
OK - I now have bazel 0.7.0 installed (from Arch User Repository - latest they have). plaidml builds fine with Python 2.7 but fails to build with Python 3.6:

```
bazel version
Build label: 0.7.0- (@non-git)
Build target: bazel-out/local-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Fri Oct 20 11:32:11 2017 (1508499131)
Build timestamp: 1508499131
Build timestamp as int: 1508499131
```

```
____Loading package: plaidml/keras
____Loading package: @bazel_tools//tools/cpp
____Loading package: @bazel_tools//tools/jdk
____Loading package: @local_config_xcode//
____Loading package: @local_jdk//
____Loading package: @local_config_cc//
____Loading complete.  Analyzing...
____Loading package: bzl
____Loading package: @bazel_tools//tools/python
____Loading package: @bazel_tools//tools/def_parser
____Loading package: @half_repo//
____Cloning https://github.com/google/protobuf: remote: Counting objects (0 / 0)
____Cloning https://github.com/google/protobuf: Receiving objects (22 / 47728)
____Cloning https://github.com/google/protobuf: Receiving objects (96 / 47728)

[snip]

____Loading package: @easylogging_repo//
____Downloading https://storage.googleapis.com/external_build_repo/boost_1_63_0.tar.bz2: 74,865,521 bytes
____Downloading https://storage.googleapis.com/external_build_repo/boost_1_63_0.tar.bz2: 76,979,057 bytes
____Downloading https://storage.googleapis.com/external_build_repo/boost_1_63_0.tar.bz2: 79,108,977 bytes
____Downloading https://storage.googleapis.com/external_build_repo/boost_1_63_0.tar.bz2: 81,230,705 bytes
____Loading package: @boost_archive//
____Loading package: @gmock_archive//
____Loading package: @bazel_tools//tools/launcher
____Found 2 targets...
____Building...
____[6 / 19] CopySetupCfg plaidml/wheel.pkg/setup.cfg
____[20 / 93] Creating runfiles tree bazel-out/host/bin/external/protobuf/js_embed.runfiles [for host]
____[28 / 172] Compiling external/zlib_archive/zlib-1.2.8/inffast.c [for host]
____From BuildWheel plaidml/keras/wheel.pkg/dist/plaidml_keras-0.0.0.dev0-py2-none-any.whl:
running bdist_wheel
running build
running build_py
creating build
creating build/lib
creating build/lib/plaidml
creating build/lib/plaidml/keras
copying pkg/plaidml/keras/__init__.py -> build/lib/plaidml/keras
copying pkg/plaidml/keras/backend.py -> build/lib/plaidml/keras
installing to build/bdist.linux-x86_64/wheel
running install
running install_lib
creating build/bdist.linux-x86_64
creating build/bdist.linux-x86_64/wheel
creating build/bdist.linux-x86_64/wheel/plaidml
creating build/bdist.linux-x86_64/wheel/plaidml/keras
copying build/lib/plaidml/keras/__init__.py -> build/bdist.linux-x86_64/wheel/plaidml/keras
copying build/lib/plaidml/keras/backend.py -> build/bdist.linux-x86_64/wheel/plaidml/keras
running install_egg_info
running egg_info
creating pkg/plaidml_keras.egg-info
writing pkg/plaidml_keras.egg-info/PKG-INFO
writing dependency_links to pkg/plaidml_keras.egg-info/dependency_links.txt
writing entry points to pkg/plaidml_keras.egg-info/entry_points.txt
writing requirements to pkg/plaidml_keras.egg-info/requires.txt
writing top-level names to pkg/plaidml_keras.egg-info/top_level.txt
writing manifest file 'pkg/plaidml_keras.egg-info/SOURCES.txt'
reading manifest file 'pkg/plaidml_keras.egg-info/SOURCES.txt'
writing manifest file 'pkg/plaidml_keras.egg-info/SOURCES.txt'
Copying pkg/plaidml_keras.egg-info to build/bdist.linux-x86_64/wheel/plaidml_keras-0.0.0.dev0-py3.6.egg-info
running install_scripts
creating build/bdist.linux-x86_64/wheel/plaidml_keras-0.0.0.dev0.dist-info/WHEEL
ERROR: /home/Installers/plaidml/plaidml/keras/BUILD:67:1: output 'plaidml/keras/wheel.pkg/dist/plaidml_keras-0.0.0.dev0-py2-none-any.whl' was not created.
ERROR: /home/Installers/plaidml/plaidml/keras/BUILD:67:1: not all outputs were created or valid.
____Building complete.
____Elapsed time: 89.710s, Critical Path: 0.59s
```

**Comment by choongng** [comment]
@znmeb Thanks for digging in to this. Python 3 support is not a high priority for us but if someone contributes a patch we'll help get it integrated into our test suite to keep it working.

**Comment by znmeb** [comment]
I have a workaround for building it - change "py2" to "py3" in the two "BUILD" scripts. But I don't know enough Python / Bazel to configure it to detect Python at run time and pick the right one.

**Comment by choongng** [comment]
Once you've got it built does plaidbench run?

**Comment by znmeb** [comment]
I haven't tried it yet - I was waiting to see if it was something easy to fix on your end. I'll try it tonight.

**Comment by znmeb** [comment]
It doesn't execute. Two issues:
1. `requirements.txt` is asking for "plaidml-keras>=0.1.0a5". What pip installed is "0.0.0.dev0". So I deleted that line from `requirements.txt` and re-ran.
2. It still isn't finding what it needs:

```
$ python plaidbench.py mobilenet 2>&1 | tee $HOME/pltest.log
Traceback (most recent call last):
  File "plaidbench.py", line 208, in <module>
    main()
  File "plaidbench.py", line 91, in main
    if args.plaid or (not args.no_plaid and has_plaid()):
  File "plaidbench.py", line 68, in has_plaid
    import plaidml.keras
  File "/home/znmeb/.conda/envs/plaidml-github/lib/python3.6/site-packages/plaidml/__init__.py", line 22, in <module>
    import plaidml.settings
  File "/home/znmeb/.conda/envs/plaidml-github/lib/python3.6/site-packages/plaidml/settings.py", line 147, in <module>
    module = _Settings()
  File "/home/znmeb/.conda/envs/plaidml-github/lib/python3.6/site-packages/plaidml/settings.py", line 40, in __init__
    self._load()
  File "/home/znmeb/.conda/envs/plaidml-github/lib/python3.6/site-packages/plaidml/settings.py", line 69, in _load
    for k, val in json.load(file(settings_file)).items():
NameError: name 'file' is not defined
 ```

**Comment by brianretford** [comment]
Not in front of a computer, but I think this because python 3 doesn't
support 'file' and instead requires 'open':
https://stackoverflow.com/questions/112970/python-when-to-use-file-vs-open

**Comment by znmeb** [comment]
@brianretford Probably - I only know enough Python to copy / paste code into Jupyter notebooks. ;-) I'll try the "open" and see if it runs.

**Comment by znmeb** [comment]
Yeah - changing `file` to `open` works. Then it crashes because it doesn't have the `tensorflow` dependency. So I installed 1.3.0 via `conda`. Now it's looking for the `tensorflow` back end instead of `plaidml`.

```
$ python plaidbench.py mobilenet 2>&1 | tee $HOME/pltest.log
Using TensorFlow backend.
Using PlaidML backend.
name 'execfile' is not defined
Set --print-stacktraces to see the entire traceback
```

**Comment by Juanlu001** [comment]
Running futurize does the trick for the workbench, see http://python-future.org/automatic_conversion.html

```
$ futurize -w --stage2 plaidbench.py
[...]
```

However, more things need fixing.

**Comment by Juanlu001** [comment]
My attempt: https://github.com/plaidml/plaidml/pull/25

**Comment by brianretford** [comment]
@Juanlu001 seems to have gotten most of the way there. The key issue now is in [plaidml/keras/__init__.py](https://github.com/plaidml/plaidml/blob/master/plaidml/keras/__init__.py#L60), specifically in `install_backend`.

We have to monkey patch our backend into keras (hopefully we can upstream it soon). The mechanism we use doesn't work in python 3. Would love help on this one.

**Comment by Juanlu001** [comment]
Hmm now that you mention it, I noticed that when running plaidbench. Will try to find a solution.

**Comment by Juanlu001** [comment]
The import machinery was rewritten in Python 3.5 IIRC, so that can be the source of the problem. Anyway, I asked for a better integration method in the keras issue linked above.

**Comment by brianretford** [comment]
Note that I had to move to six instead of future because of: https://github.com/PythonCharmers/python-future/pull/307 (in the current update from vertex.ai)

**Comment by sw1sh** [comment]
Trying to make `plaidml-setup` work: #42

**Comment by brianretford** [comment]
Just about got this wrapped up here. Thanks everyone!

**Comment by brianretford** [comment]
Resolved in #45
